oeps/oep-0022-bp-django-caches.rst

```diff
@@ -0,0 +1,229 @@
+        # if the calculated value is Falsey (e.g., None or []), it will
+        # be treated as a cache miss above.
+    return cached_value
+
+The simplest fix for this defect is::
+
+    _CACHE_MISS = object()
+
+    cached_value = cache.get(key, _CACHE_MISS)
+    if cached_value is _CACHE_MISS:
+        # calculate value, set in cache, and return
+        # if the calculated value is Falsey (e.g., None or []), it will
+        # no longer be treated as a cache miss above.
+    return cached_value
+
+Use libraries that makes it simpler to avoid the above defect. See
+`Implementations`_ for more details.
+
+Tiered Cache with Request Caching
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use a library that allows tiered-caching calls that use a Request Cache backed
+by Memcached. Multiple calls for the same value within a request will not
+require additional network calls to Memcached.
+
+See `Implementations`_ for more details.
+
+Force Cache Refresh
+~~~~~~~~~~~~~~~~~~~
+
+Ensure that new caching functionality has a simple method by which to force a
+cache refresh. This is useful both for testing purposes, as well as for
+business purposes if a problem occurs in Production that requires a capability
+for refreshing a cached value.
+
+Cache Decorators
+~~~~~~~~~~~~~~~~
+
+The following recommendation requires additional assessment.
+
+When caching, create cache methods where the boilerplate caching code is hidden
+behind a decorator. This practice allows the more important business logic to
+stand out.
+
+Note that the boilerplate logic hidden behind the decorator must include
+opinionated code regarding how keys are defined and when and if the key will
+change. More assessment is required here to determine the recommended
+decorator(s) to use.
+
+Sample Before::
+
+    def get_and_cache_my_business_value(x):
+        cached_value = cache.get(key, _CACHE_MISS)
+        if cached_value is _CACHE_MISS:
+            new_value = calculate_my_business_value()
+            key = get_my_business_key(x)
+            cache.set(key, new_value)
+            return new_value
+        return cached_value
+
+Sample After::
+
+    @sample_cache_decorator(['x'])
+    def get_and_cache_my_business_value(x):
+        # calculate my_business_value
+        ...
+        return my_business_value
+
+See `Implementations`_ for more details.
+
+Generating Cache Keys
+~~~~~~~~~~~~~~~~~~~~~
+
+Use a utility function that will provide a Memcached safe key that will avoid
+key conflicts in the global space in Memcached. This functionality may or may
+not come as part of a caching decorator.
+
+See `Implementations`_ for more details.
+
+Implementations
+~~~~~~~~~~~~~~~
+
+This section will be updated as a shared library is selected and evolved.
+
+**In edx-platform:**
+
+* There is no utility for handling the `Common Caching Defect and Fix`_.
+
+* There is an `edx-platform RequestCache`_, but no Tiered-Cache.
+
+* There are no utilities for forcing cache misses.
```

Comment:
To force cache misses, we clear the cache before calling the method that uses and populates the cache.